### ISSUE TITLE
offline-installer: still uses JMX and get tested with old nodetool

### ISSUE THREAD

**Issue by fruch** [scylladb/scylladb#19185]
While trying to introduce a new distro into SCT, I've notice that the offline installer tests are still using the old nodetools, and
it even works most of the time, 

since the unified installer still comes with JMX and nodetool.

shouldn't we remove the JMX/nodetool out of unified package as well ?

**Comment by mykaul** [comment]
Does it get installed, or also gets enabled? I think the former makes sense, the latter - less so.

**Comment by fruch** [comment]
> Does it get installed, or also gets enabled? I think the former makes sense, the latter - less so.

both, it is installed, and being used. (i.e. the tests never go changed to use the new tool, and there still pending issue, that user need to configure )

**Comment by mykaul** [comment]
> > Does it get installed, or also gets enabled? I think the former makes sense, the latter - less so.
> 
> both, it is installed, and being used. (i.e. the tests never go changed to use the new tool, and there still pending issue, that user need to configure )

Not a very high priority IMHO, we can change it. I do wonder why the default upon installation is to enable it. We probably should switch it to default disabled.

**Comment by fruch** [comment]
it's hard coded in the test, to use the old version of the tool.

and when changing it, we can get into:
https://github.com/scylladb/scylladb/issues/17412

i.e. /proc/sys/fs/aio-max-nr isn't big enough, and user should manual change it (i.e. not documented)

**Comment by mykaul** [comment]
@fruch - so the issue is to change the tests?

**Comment by fruch** [comment]
> @fruch - so the issue is to change the tests?

No, the issue is to remove the old tool from unified package (and yes at the same time the test would need to be adapted)

**Comment by mykaul** [comment]
> > @fruch - so the issue is to change the tests?
> 
> No, the issue is to remove the old tool from unified package (and yes at the same time the test would need to be adapted)

Since the tool is optional, I feel it should be installed as part of offline installation, but disabled by default. Perhaps this should be its default everywhere.

**Comment by fruch** [comment]
> > > @fruch - so the issue is to change the tests?
> > 
> > 
> > No, the issue is to remove the old tool from unified package (and yes at the same time the test would need to be adapted)
> 
> Since the tool is optional, I feel it should be installed as part of offline installation, but disabled by default. Perhaps this should be its default everywhere.

JMX isn't optional now in offline installer, it's being installed, and JMX is running, regardless if you use the java base nodetool or not.

**Comment by denesb** [comment]
We have https://github.com/scylladb/scylladb/pull/17969 to drop JMX from packaging.

**Comment by denesb** [comment]
> it's hard coded in the test, to use the old version of the tool.
> 
> and when changing it, we can get into: #17412
> 
> i.e. /proc/sys/fs/aio-max-nr isn't big enough, and user should manual change it (i.e. not documented)

I though this was solved. The aio-max-nr was solved by changing the reactor backend a long time ago, then there was another unrelated failure which was fixed as well.
AFAIK the latest runs passed.

**Comment by mykaul** [comment]
Unsure what's the next step here?

**Comment by denesb** [comment]
I thought we already removed jmx from the offline installer, as part of https://github.com/scylladb/scylladb/pull/18487.

**Comment by fruch** [comment]
> I thought we already removed jmx from the offline installer, as part of #18487.

@denesb that PR only disabled the installation of it, JMX just recently was removed from unified package,

and now all of the offline install non-root cases, are failing to run nodetool:
```
< t:2024-09-22 18:08:32,206 f:remote_base.py  l:560  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.34>: Running command "/home/scylla-test/scylladb/bin/nodetool  status "...
< t:2024-09-22 18:08:32,709 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.34>: terminate called after throwing an instance of 'std::runtime_error'
< t:2024-09-22 18:08:33,209 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.34>:   what():  Could not setup Async I/O: Resource temporarily unavailable. The required nr_event (1) exceeds the limit of request capacity in /proc/sys/fs/aio-max-nr (65536). Try increasing that number or reducing the amount of logical CPUs available for your application
< t:2024-09-22 18:08:33,210 f:base.py         l:141  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.34>: STDERR: terminate called after throwing an instance of 'std::runtime_error'
< t:2024-09-22 18:08:33,210 f:base.py         l:141  c:RemoteLibSSH2CmdRunner p:DEBUG >   what():  Could not setup Async I/O: Resource temporarily unavailable. The required nr_event (1) exceeds the limit of request capacity in /proc/sys/fs/aio-max-nr (65536). Try increasing that number or reducing the amount of logical CPUs available for your application
< t:2024-09-22 18:08:33,210 f:base.py         l:143  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.34>: Command "/home/scylla-test/scylladb/bin/nodetool  status " finished with status 0
< t:2024-09-22 18:08:33,210 f:cluster.py      l:2664 c:sdcm.cluster_gce     p:DEBUG > Node artifacts-ubuntu2204-jenkins-db-node-619fb589-0-1 [34.23.18.68 | 10.142.0.34]: Command '/home/scylla-test/scylladb/bin/nodetool  status ' duration -> 1.0030242060000774 s
```

**Comment by fruch** [comment]
PR switch to the new node tool:
- https://github.com/scylladb/scylla-cluster-tests/pull/8788

**Comment by yaronkaikov** [comment]
> > I thought we already removed jmx from the offline installer, as part of #18487.
> 
> @denesb that PR only disabled the installation of it, JMX just recently was removed from unified package,
> 
> and now all of the offline install non-root cases, are failing to run nodetool:
> 
> ```
> < t:2024-09-22 18:08:32,206 f:remote_base.py  l:560  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.34>: Running command "/home/scylla-test/scylladb/bin/nodetool  status "...
> < t:2024-09-22 18:08:32,709 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.34>: terminate called after throwing an instance of 'std::runtime_error'
> < t:2024-09-22 18:08:33,209 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.34>:   what():  Could not setup Async I/O: Resource temporarily unavailable. The required nr_event (1) exceeds the limit of request capacity in /proc/sys/fs/aio-max-nr (65536). Try increasing that number or reducing the amount of logical CPUs available for your application
> < t:2024-09-22 18:08:33,210 f:base.py         l:141  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.34>: STDERR: terminate called after throwing an instance of 'std::runtime_error'
> < t:2024-09-22 18:08:33,210 f:base.py         l:141  c:RemoteLibSSH2CmdRunner p:DEBUG >   what():  Could not setup Async I/O: Resource temporarily unavailable. The required nr_event (1) exceeds the limit of request capacity in /proc/sys/fs/aio-max-nr (65536). Try increasing that number or reducing the amount of logical CPUs available for your application
> < t:2024-09-22 18:08:33,210 f:base.py         l:143  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.34>: Command "/home/scylla-test/scylladb/bin/nodetool  status " finished with status 0
> < t:2024-09-22 18:08:33,210 f:cluster.py      l:2664 c:sdcm.cluster_gce     p:DEBUG > Node artifacts-ubuntu2204-jenkins-db-node-619fb589-0-1 [34.23.18.68 | 10.142.0.34]: Command '/home/scylla-test/scylladb/bin/nodetool  status ' duration -> 1.0030242060000774 s
> ```

@fruch do you know if before this failed command Scylla is running? where can i find the logs ?

**Comment by denesb** [comment]
> The required nr_event (1) exceeds the limit of request capacity in /proc/sys/fs/aio-max-nr (65536)

This message makes sense. The current limit is `65536`, yet nodetool fails to obtain its `1` (!) unit. I think the only explanation is that all `65536` units are taken by a running ScyllaDB instance. Just bump this limit by at least `1` and it should solve the issue.

**Comment by fruch** [comment]
> > > I thought we already removed jmx from the offline installer, as part of #18487.
> > 
> > @denesb that PR only disabled the installation of it, JMX just recently was removed from unified package,
> > 
> > and now all of the offline install non-root cases, are failing to run nodetool:
> > 
> > ```
> > < t:2024-09-22 18:08:32,206 f:remote_base.py  l:560  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.34>: Running command "/home/scylla-test/scylladb/bin/nodetool  status "...
> > < t:2024-09-22 18:08:32,709 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.34>: terminate called after throwing an instance of 'std::runtime_error'
> > < t:2024-09-22 18:08:33,209 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.34>:   what():  Could not setup Async I/O: Resource temporarily unavailable. The required nr_event (1) exceeds the limit of request capacity in /proc/sys/fs/aio-max-nr (65536). Try increasing that number or reducing the amount of logical CPUs available for your application
> > < t:2024-09-22 18:08:33,210 f:base.py         l:141  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.34>: STDERR: terminate called after throwing an instance of 'std::runtime_error'
> > < t:2024-09-22 18:08:33,210 f:base.py         l:141  c:RemoteLibSSH2CmdRunner p:DEBUG >   what():  Could not setup Async I/O: Resource temporarily unavailable. The required nr_event (1) exceeds the limit of request capacity in /proc/sys/fs/aio-max-nr (65536). Try increasing that number or reducing the amount of logical CPUs available for your application
> > < t:2024-09-22 18:08:33,210 f:base.py         l:143  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.34>: Command "/home/scylla-test/scylladb/bin/nodetool  status " finished with status 0
> > < t:2024-09-22 18:08:33,210 f:cluster.py      l:2664 c:sdcm.cluster_gce     p:DEBUG > Node artifacts-ubuntu2204-jenkins-db-node-619fb589-0-1 [34.23.18.68 | 10.142.0.34]: Command '/home/scylla-test/scylladb/bin/nodetool  status ' duration -> 1.0030242060000774 s
> > ```
> 
> @fruch do you know if before this failed command Scylla is running? where can i find the logs ?

Yes scylla should be running, no point of doing any nodetool commands if Scylla isn't up. 

But in some situations we might run it when scylla isn't up, but we retry until it would work. which doesn't happen in those runs

**Comment by denesb** [comment]
> > > The required nr_event (1) exceeds the limit of request capacity in /proc/sys/fs/aio-max-nr (65536)
> > 
> > 
> > This message makes sense. The current limit is `65536`, yet nodetool fails to obtain its `1` (!) unit. I think the only explanation is that all `65536` units are taken by a running ScyllaDB instance. Just bump this limit by at least `1` and it should solve the issue.
> 
> Kind of contradicts a non-root installation, if the user needs root access to tune anything.

You need this to run non-root scylla in the first place. Some tuning is necessary.

> Anyhow it's a regression, since the java nodetool didn't need that tweak.
> 
> This kind of behavior change should be documented and approved. (Or worked around in other means)

I tried to work around it by using the epoll backend instead of the AIO one. I don't understand why seastar still tries to allocate a unit. I will look into it.

**Comment by denesb** [comment]
Ok, I think I found a bug in seastar related to this.

**Comment by denesb** [comment]
seastar bug which causes this failing check: https://github.com/scylladb/seastar/issues/2453.

**Comment by fruch** [comment]
> seastar bug which causes this failing check: [scylladb/seastar#2453](https://github.com/scylladb/seastar/issues/2453).

@denesb this didn't solved the issue,
see this run:

https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-ubuntu2404-test/16/

it's running scylla `6.3.0~dev-0.20241004.882a3c60e4a5` but still failing to run nodetool:

```
< t:2024-10-06 10:00:32,462 f:remote_base.py  l:560  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.61>: Running command "/home/scylla-test/scylladb/bin/nodetool  status "...
< t:2024-10-06 10:00:32,964 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.61>: terminate called after throwing an instance of 'std::runtime_error'
< t:2024-10-06 10:00:33,464 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.61>:   what():  Could not setup Async I/O: Resource temporarily unavailable. The required nr_event (1) exceeds the limit of request capacity in /proc/sys/fs/aio-max-nr (65536). Try increasing that number or reducing the amount of logical CPUs available for your application
```

**Comment by yaronkaikov** [comment]
> > seastar bug which causes this failing check: [scylladb/seastar#2453](https://github.com/scylladb/seastar/issues/2453).
> 
> [@denesb](https://github.com/denesb) this didn't solved the issue, see this run:
> 
> https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/artifacts-ubuntu2404-test/16/
> 
> it's running scylla `6.3.0~dev-0.20241004.882a3c60e4a5` but still failing to run nodetool:
> 
> ```
> < t:2024-10-06 10:00:32,462 f:remote_base.py  l:560  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.61>: Running command "/home/scylla-test/scylladb/bin/nodetool  status "...
> < t:2024-10-06 10:00:32,964 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.61>: terminate called after throwing an instance of 'std::runtime_error'
> < t:2024-10-06 10:00:33,464 f:base.py         l:231  c:RemoteLibSSH2CmdRunner p:DEBUG > <10.142.0.61>:   what():  Could not setup Async I/O: Resource temporarily unavailable. The required nr_event (1) exceeds the limit of request capacity in /proc/sys/fs/aio-max-nr (65536). Try increasing that number or reducing the amount of logical CPUs available for your application
> ```

@denesb Any idea?

**Comment by denesb** [comment]
> [@denesb](https://github.com/denesb) Any idea?

No idea.

Can you add `--logger-log-level seastar=trace` to the `nodetool` command line, so we can see what it is doing?

**Comment by yaronkaikov** [comment]
> > [@denesb](https://github.com/denesb) Any idea?
> 
> No idea.
> 
> Can you add `--logger-log-level seastar=trace` to the `nodetool` command line, so we can see what it is doing?

@fruch Can you please create some staging branch for this?

**Comment by yaronkaikov** [comment]
> > > [@denesb](https://github.com/denesb) Any idea?
> > 
> > 
> > No idea.
> > Can you add `--logger-log-level seastar=trace` to the `nodetool` command line, so we can see what it is doing?
> 
> [@fruch](https://github.com/fruch) Can you please create some staging branch for this?

/cc @mykaul

**Comment by fruch** [comment]
@denesb 

`/home/scylla-test/scylladb/bin/nodetool --logger-log-level seastar=trace  status` doesn't seem to work...
I have no idea how to pass this into the tool.

I would suggest someone would run the staging job mentioned above with keep, and debug it on the actual machines.

**Comment by denesb** [comment]
> [@denesb](https://github.com/denesb)
> 
> `/home/scylla-test/scylladb/bin/nodetool --logger-log-level seastar=trace status` doesn't seem to work... I have no idea how to pass this into the tool.
> 
> I would suggest someone would run the staging job mentioned above with keep, and debug it on the actual machines.

Works for me, I have no environment at hand to reproduce this at, hence the many fruitless back-and-forth.

**Comment by mykaul** [comment]
@fruch , @denesb - what's the next step here then?

**Comment by denesb** [comment]
> [@fruch](https://github.com/fruch) , [@denesb](https://github.com/denesb) - what's the next step here then?

I'm waiting for somebody to either give me logs with the increased logging or access to a machine where I can reproduce this. It doesn't reproduce locally.

**Comment by dani-tweig** [comment]
> I'm waiting for somebody to either give me logs with the increased logging or access to a machine where I can reproduce this. It doesn't reproduce locally.

@fruch , @pehala , who can help @denesb ?

**Comment by denesb** [comment]
`reactor_backend_epoll` also creates an `aio_storage_context` which also requires AIO units.  Looks like there is no way around this. Maybe we even have to revert https://github.com/scylladb/seastar/commit/3c9c2696a45ec0eea524760a75938900d75f79b8 as the check was justified after all.

@avikivity is there a way to avoid requiring AIO units for tools, or we have to bite the bullet and require that all nodes have units for Scylla + minimum reserve for a tool.

**Comment by avikivity** [comment]
> [@avikivity](https://github.com/avikivity) is there a way to avoid requiring AIO units for tools, or we have to bite the bullet and require that all nodes have units for Scylla + minimum reserve for a tool.

We can add a way to prefer epoll/io_uring in reactor_config, then set the tools to prefer those backends.

Provided we know we're a tool ahead of launching the reactor.

**Comment by denesb** [comment]
> > [@avikivity](https://github.com/avikivity) is there a way to avoid requiring AIO units for tools, or we have to bite the bullet and require that all nodes have units for Scylla + minimum reserve for a tool.
> 
> We can add a way to prefer epoll/io_uring in reactor_config, then set the tools to prefer those backends.
> 
> Provided we know we're a tool ahead of launching the reactor.

The tools are already configured to use the epoll backend, but the epoll backend relies on AIO blocks via the `aio_storage_context`, see https://github.com/scylladb/seastar/blob/fba36a3d15e22a112b6eedf40f653560ca574274/src/core/reactor_backend.hh#L246.

**Comment by avikivity** [comment]
> > > [@avikivity](https://github.com/avikivity) is there a way to avoid requiring AIO units for tools, or we have to bite the bullet and require that all nodes have units for Scylla + minimum reserve for a tool.
> > 
> > We can add a way to prefer epoll/io_uring in reactor_config, then set the tools to prefer those backends.
> > 
> > Provided we know we're a tool ahead of launching the reactor.
> 
> The tools are already configured to use the epoll backend, but the epoll backend relies on AIO blocks via the `aio_storage_context`, see https://github.com/scylladb/seastar/blob/fba36a3d15e22a112b6eedf40f653560ca574274/src/core/reactor_backend.hh#L246.

We could add a thread pool backend for IO. Or enable io_uring and let the application configure the choices at runtime.

**Comment by denesb** [comment]
> We could add a thread pool backend for IO. Or enable io_uring and let the application configure the choices at runtime.

I think this should work already, so long as the io_uring is compiled in. Is the io_uring backend feature complete?

**Comment by denesb** [comment]
@avikivity the io_uring backend also calls `setup_aio_context()`: preempt_io_context() -> `general_aio_context()` -> `setup_aio_context()`.

**Comment by denesb** [comment]
In conclusion: to make these tests pass, we need to increase the amount of IO blocks on the server, to accommodate nodetool too.
@fruch @pehala

**Comment by fruch** [comment]
> In conclusion: to make these tests pass, we need to increase the amount of IO blocks on the server, to accommodate nodetool too. [@fruch](https://github.com/fruch) [@pehala](https://github.com/pehala)

first it should be documented in the instruction for offline installer, and then we can align the test code to it.

**Comment by mykaul** [comment]
> > In conclusion: to make these tests pass, we need to increase the amount of IO blocks on the server, to accommodate nodetool too. [@fruch](https://github.com/fruch) [@pehala](https://github.com/pehala)
> 
> first it should be documented in the instruction for offline installer, and then we can align the test code to it.

Perhaps it's worth verifying it works first? Why are we setting conditions to testing it?

**Comment by fruch** [comment]
> > > In conclusion: to make these tests pass, we need to increase the amount of IO blocks on the server, to accommodate nodetool too. [@fruch](https://github.com/fruch) [@pehala](https://github.com/pehala)
> > 
> > 
> > first it should be documented in the instruction for offline installer, and then we can align the test code to it.
> 
> Perhaps it's worth verifying it works first? Why are we setting conditions to testing it?

cause we are testing base on docs, and those test are following the instruction to the letter.

also changing a kernel option contradicts a user that doesn't have root access, which is the case we are talking about.

I'm not gonna change the test until we clarify it, someone want to try it manually, be my guest.

**Comment by avikivity** [comment]
> [@avikivity](https://github.com/avikivity) the io_uring backend also calls `setup_aio_context()`: `preempt_io_context()` -> `general_aio_context()` -> `setup_aio_context()`.

Yes, I forgot. I racked my brains about how to get an async write to user memory using io_uring, but disn't find a solution.

**Comment by avikivity** [comment]
> > [@avikivity](https://github.com/avikivity) the io_uring backend also calls `setup_aio_context()`: `preempt_io_context()` -> `general_aio_context()` -> `setup_aio_context()`.
> 
> Yes, I forgot. I racked my brains about how to get an async write to user memory using io_uring, but disn't find a solution.

But, it needs a truly tiny number of aio contexts.

**Comment by denesb** [comment]
> > > [@avikivity](https://github.com/avikivity) the io_uring backend also calls `setup_aio_context()`: `preempt_io_context()` -> `general_aio_context()` -> `setup_aio_context()`.
> > 
> > 
> > Yes, I forgot. I racked my brains about how to get an async write to user memory using io_uring, but disn't find a solution.
> 
> But, it needs a truly tiny number of aio contexts.

Yes, it needs 2 units. We can surely spare that. I just need to find a way to tell ScyllaDB (seastar) to leave some units aside as spare for nodetool.

**Comment by denesb** [comment]
I have a working patch, it involves:
* adding a iocb reserve config to seastar
* setting it in scylla's main.cc to 10 (strictly speaking 2 is enough)
* change configure.py to include io_uring reactor backend
* use io_uring reactor backend in tools

Alternatively, we can keep using the epoll backend, but then we need a reserve of 1024 iocb. While I feel like we won't feel 10 iocb missing, we might feel 1024 iocb missing (but this is just me guessing).

**Comment by denesb** [comment]
I have verified and it works with epoll too, as long as the reserve is sufficiently large (1024).
So the `io_uring` option means we get away with smaller reserve (as small as 2), but on the down side, we enable the unstable `io_uring` backend which is then chosen by default by seastar (can be worked around in `main.cc`) and it can be chosen by the user.

@avikivity which one should we go for? Stay with `epoll` and a reserve of 1024 or go for `io_uring` at the cost of some extra work (and maybe a bit of extra bit on the account of the less table backend)?

**Comment by avikivity** [comment]
Is io_uring unstable? I thought the only problem was performance.

If it's just performance, we can make the backend priority run-time configurable (provided we know we're a tool ahead of starting Seastar). It would help getting mileage into io_uring.

**Comment by denesb** [comment]
> Is io_uring unstable? I thought the only problem was performance.

I don't know, I'm just quoting comments I've seen in the code, in particular in our `configure.py`:
https://github.com/scylladb/scylladb/blob/6efde20939d072406b6e80a7ece1026134b0268a/configure.py#L1708

> If it's just performance, we can make the backend priority run-time configurable (provided we know we're a tool ahead of starting Seastar). It would help getting mileage into io_uring.

We already do a tool-oriented config when starting up tools:
https://github.com/scylladb/scylladb/blob/6efde20939d072406b6e80a7ece1026134b0268a/tools/utils.cc#L140-L163

**Comment by avikivity** [comment]
> > Is io_uring unstable? I thought the only problem was performance.
> 
> I don't know, I'm just quoting comments I've seen in the code, in particular in our `configure.py`:
> 
> [scylladb/configure.py](https://github.com/scylladb/scylladb/blob/6efde20939d072406b6e80a7ece1026134b0268a/configure.py#L1708)
> 
> Line 1708 in [6efde20](/scylladb/scylladb/commit/6efde20939d072406b6e80a7ece1026134b0268a)
> 
>  '-DSeastar_IO_URING=OFF', # io_uring backend is not stable enough 
> > If it's just performance, we can make the backend priority run-time configurable (provided we know we're a tool ahead of starting Seastar). It would help getting mileage into io_uring.
> 

I remember that it's only performance. In any case, it will never get stable without us using it, so let's try it in tools.


> We already do a tool-oriented config when starting up tools:
> 
> [scylladb/tools/utils.cc](https://github.com/scylladb/scylladb/blob/6efde20939d072406b6e80a7ece1026134b0268a/tools/utils.cc#L140-L163)
> 
> Lines 140 to 163 in [6efde20](/scylladb/scylladb/commit/6efde20939d072406b6e80a7ece1026134b0268a)
> 
>  // Configure seastar with defaults more appropriate for a tool. 
>  // Make seastar not act as if it owns the place, taking over all system resources. 
>  // Set ERROR as the default log level, except for the logger \p logger_name, which 
>  // is configured with INFO level. 
>  void configure_tool_mode(app_template::seastar_options& opts, const sstring& logger_name) { 
>      opts.reactor_opts.blocked_reactor_notify_ms.set_value(60000); 
>      opts.reactor_opts.overprovisioned.set_value(); 
>      opts.reactor_opts.idle_poll_time_us.set_value(0); 
>      opts.reactor_opts.poll_aio.set_value(false); 
>      opts.reactor_opts.relaxed_dma.set_value(); 
>      opts.reactor_opts.unsafe_bypass_fsync.set_value(true); 
>      opts.reactor_opts.kernel_page_cache.set_value(true); 
>      opts.reactor_opts.reactor_backend.select_candidate("epoll"); 
>      opts.smp_opts.thread_affinity.set_value(false); 
>      opts.smp_opts.mbind.set_value(false); 
>      opts.smp_opts.smp.set_value(1); 
>      opts.smp_opts.lock_memory.set_value(false); 
>      opts.smp_opts.memory_allocator = memory_allocator::standard; 
>      opts.log_opts.default_log_level.set_value(log_level::error); 
>      if (!logger_name.empty()) { 
>          opts.log_opts.logger_log_level.set_value({}); 
>          opts.log_opts.logger_log_level.get_value()[logger_name] = log_level::info; 
>      } 
>  }


We can add a vector for backend priority here.

**Comment by denesb** [comment]
> I remember that it's only performance. In any case, it will never get stable without us using it, so let's try it in tools.

Ok, I will prepare the patches.

**Comment by denesb** [comment]
Fix here: https://github.com/scylladb/scylladb/pull/21527
seastar part: https://github.com/scylladb/seastar/pull/2533

**Comment by yaronkaikov** [comment]
> Fix here: [#21527](https://github.com/scylladb/scylladb/pull/21527) seastar part: [scylladb/seastar#2533](https://github.com/scylladb/seastar/pull/2533)

@denesb i see https://github.com/scylladb/seastar/pull/2533 already merged, did we update the submodule ?
